<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quick start to Harmony • harmony</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Quick start to Harmony">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">harmony</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../advanced.html">Detailed Walkthrough</a>
    </li>
    <li>
      <a href="../mudan.html">MUDAN</a>
    </li>
    <li>
      <a href="../SeuratV2.html">Seurat V2</a>
    </li>
    <li>
      <a href="../SeuratV3.html">Seurat V3</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/immunogenomics/harmony">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Quick start to Harmony</h1>
                        <h4 class="author">Korsunsky et al.: Fast, sensitive, and accurate integration of single cell data with Harmony</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/immunogenomics/harmony/blob/master/vignettes/quickstart.Rmd"><code>vignettes/quickstart.Rmd</code></a></small>
      <div class="hidden name"><code>quickstart.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Harmony is an algorithm for performing integration of single cell genomics datasets. Please check out our latest <a href="https://www.biorxiv.org/content/10.1101/461954v2">preprint on bioRxiv</a>.</p>
<p><img src="main.jpg"></p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>Install Harmony with standard Bioconductor commands.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">    <span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"harmony"</span>, <span class="dt">version =</span> <span class="st">"3.8"</span>)</a></code></pre></div>
</div>
<div id="integrating-cell-line-datasets-from-10x" class="section level1">
<h1 class="hasAnchor">
<a href="#integrating-cell-line-datasets-from-10x" class="anchor"></a>Integrating cell line datasets from 10X</h1>
<p>The example below follows Figure 2 in the manuscript.</p>
<p>We downloaded 3 cell line datasets from the 10X website. The first two (jurkat and 293t) come from pure cell lines while the <em>half</em> dataset is a 50:50 mixture of Jurkat and HEK293T cells. We inferred cell type with the canonical marker XIST, since the two cell lines come from 1 male and 1 female donor.</p>
<ul>
<li>support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/jurkat</li>
<li>support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/293t</li>
<li>support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/jurkat:293t_50:50</li>
</ul>
<p>We library normalized the cells, log transformed the counts, and scaled the genes. Then we performed PCA and kept the top 20 PCs. The PCA embeddings and meta data are available as part of this package.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(cell_lines)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">V &lt;-<span class="st"> </span>cell_lines<span class="op">$</span>scaled_pcs</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">meta_data &lt;-<span class="st"> </span>cell_lines<span class="op">$</span>meta_data</a></code></pre></div>
<p>Initially, the cells cluster by both dataset (left) and cell type (right).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">p1 &lt;-<span class="st"> </span><span class="kw">do_scatter</span>(V, meta_data, <span class="st">'dataset'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">'Colored by dataset'</span>)</a></code></pre></div>
<pre><code>## Warning: group_by_() is deprecated. 
## Please use group_by() instead
## 
## The 'programming' vignette or the tidyeval book can help you
## to program with group_by() : https://tidyeval.tidyverse.org
## This warning is displayed once per session.</code></pre>
<pre><code>## Warning: Ignoring unknown parameters: segment.size</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">p2 &lt;-<span class="st"> </span><span class="kw">do_scatter</span>(V, meta_data, <span class="st">'cell_type'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">'Colored by cell type'</span>)</a></code></pre></div>
<pre><code>## Warning: Ignoring unknown parameters: segment.size</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(p1, p2)</a></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Let’s run Harmony to remove the influence of dataset-of-origin from the embedding. By default, Harmony accepts a normalized gene expression matrix and performs PCA. Since here we already have the PCs, we specify <em>do_pca=FALSE</em>. The matrix <em>harmony_embeddings</em> is the matrix of Harmony corrected PCA embeddings.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">harmony_embeddings &lt;-<span class="st"> </span>harmony<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/harmony/man/HarmonyMatrix.html">HarmonyMatrix</a></span>(</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">    V, meta_data, <span class="st">'dataset'</span>, <span class="dt">do_pca =</span> <span class="ot">FALSE</span>, <span class="dt">verbose=</span><span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">)</a></code></pre></div>
<p>After Harmony, the datasets are now mixed (left) and the cell types are still separate (right).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">p1 &lt;-<span class="st"> </span><span class="kw">do_scatter</span>(harmony_embeddings, meta_data, <span class="st">'dataset'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">'Colored by dataset'</span>)</a></code></pre></div>
<pre><code>## Warning: Ignoring unknown parameters: segment.size</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">p2 &lt;-<span class="st"> </span><span class="kw">do_scatter</span>(harmony_embeddings, meta_data, <span class="st">'cell_type'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">'Colored by cell type'</span>)</a></code></pre></div>
<pre><code>## Warning: Ignoring unknown parameters: segment.size</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(p1, p2, <span class="dt">nrow =</span> <span class="dv">1</span>)</a></code></pre></div>
<p><img src="quickstart_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div id="next-steps" class="section level1">
<h1 class="hasAnchor">
<a href="#next-steps" class="anchor"></a>Next Steps</h1>
<div id="interfacing-to-software-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#interfacing-to-software-packages" class="anchor"></a>Interfacing to software packages</h2>
<p>You can also run Harmony as part of an established pipeline in several packages, such as Seurat, MUDAN, and scran. For these vignettes, please <a href="software.broadinstitute.org/harmony">visit our website</a>.</p>
<p>The most common way to run Harmony is on reduced dimensions such as PC embeddings from principal component analysis (PCA). If you use low dimensional embeddings, set <code>do_pca = FALSE</code>. A small single-cell RNA-seq dataset is include in the Harmony package.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(harmony)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(cell_lines)</a>
<a class="sourceLine" id="cb15-4" data-line-number="4"></a>
<a class="sourceLine" id="cb15-5" data-line-number="5">cell_lines<span class="op">$</span>meta_data[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,]</a>
<a class="sourceLine" id="cb15-6" data-line-number="6">## # A tibble: 6 x 5</a>
<a class="sourceLine" id="cb15-7" data-line-number="7">##   cell_id               dataset nGene percent_mito cell_type</a>
<a class="sourceLine" id="cb15-8" data-line-number="8">##   &lt;chr&gt;                 &lt;chr&gt;   &lt;int&gt;        &lt;dbl&gt; &lt;chr&gt;</a>
<a class="sourceLine" id="cb15-9" data-line-number="9">## 1 half_GTACGAACCACCAA   half     1508       0.0148 jurkat</a>
<a class="sourceLine" id="cb15-10" data-line-number="10">## 2 t293_AGGTCATGCACTTT   t293     4009       0.0232 t293</a>
<a class="sourceLine" id="cb15-11" data-line-number="11">## 3 half_ATAGTTGACTTCTA   half     3545       0.0153 jurkat</a>
<a class="sourceLine" id="cb15-12" data-line-number="12">## 4 half_GAGCGGCTTGCTTT   half     2450       0.0170 jurkat</a>
<a class="sourceLine" id="cb15-13" data-line-number="13">## 5 jurkat_CTGATACTCCGTAA jurkat   2388       0.0601 jurkat</a>
<a class="sourceLine" id="cb15-14" data-line-number="14"></a>
<a class="sourceLine" id="cb15-15" data-line-number="15">cell_lines<span class="op">$</span>scaled_pcs[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb15-16" data-line-number="16">##              X1            X2            X3           X4           X5</a>
<a class="sourceLine" id="cb15-17" data-line-number="17">## 1:  0.002806741 -0.0014502980 -0.0063901369 0.0002824677  0.001444069</a>
<a class="sourceLine" id="cb15-18" data-line-number="18">## 2: -0.011669691  0.0008773044  0.0008972694 0.0013239566 -0.003291763</a>
<a class="sourceLine" id="cb15-19" data-line-number="19">## 3:  0.009334467 -0.0069718865 -0.0025989384 0.0018824674 -0.003803677</a>
<a class="sourceLine" id="cb15-20" data-line-number="20">## 4:  0.006339979 -0.0025175931 -0.0043900882 0.0002737979 -0.002497476</a>
<a class="sourceLine" id="cb15-21" data-line-number="21">## 5:  0.008545583  0.0070866873 -0.0022541011 0.0016793801  0.004554607</a>
<a class="sourceLine" id="cb15-22" data-line-number="22"></a>
<a class="sourceLine" id="cb15-23" data-line-number="23"><span class="co"># Run the Harmony algorithm on the PCs.</span></a>
<a class="sourceLine" id="cb15-24" data-line-number="24">harmonized_pcs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/HarmonyMatrix.html">HarmonyMatrix</a></span>(</a>
<a class="sourceLine" id="cb15-25" data-line-number="25">  <span class="dt">data_mat  =</span> cell_lines<span class="op">$</span>scaled_pcs,</a>
<a class="sourceLine" id="cb15-26" data-line-number="26">  <span class="dt">meta_data =</span> cell_lines<span class="op">$</span>meta_data,</a>
<a class="sourceLine" id="cb15-27" data-line-number="27">  <span class="dt">vars_use  =</span> <span class="st">"dataset"</span>,</a>
<a class="sourceLine" id="cb15-28" data-line-number="28">  <span class="dt">do_pca    =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb15-29" data-line-number="29">)</a>
<a class="sourceLine" id="cb15-30" data-line-number="30"></a>
<a class="sourceLine" id="cb15-31" data-line-number="31">harmonized_pcs[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb15-32" data-line-number="32">##                X1            X2            X3           X4           X5</a>
<a class="sourceLine" id="cb15-33" data-line-number="33">## [1,]  0.004328379  0.0023091631 -4.207100e-03 0.0005384881  0.001530069</a>
<a class="sourceLine" id="cb15-34" data-line-number="34">## [2,] -0.011658395  0.0001938143  3.168350e-04 0.0010923436 -0.003427295</a>
<a class="sourceLine" id="cb15-35" data-line-number="35">## [3,]  0.010934694 -0.0025131628 -6.655157e-05 0.0019817633 -0.003617146</a>
<a class="sourceLine" id="cb15-36" data-line-number="36">## [4,]  0.008157150  0.0018712240 -1.502931e-03 0.0006383834 -0.002240632</a>
<a class="sourceLine" id="cb15-37" data-line-number="37">## [5,]  0.006956255  0.0032281410 -4.582874e-03 0.0013581108  0.004490102</a></code></pre></div>
</div>
<div id="normalized-gene-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#normalized-gene-matrix" class="anchor"></a>Normalized gene matrix</h2>
<p>You can also run Harmony on a sparse matrix of library size normalized expression counts. The <code><a href="../reference/HarmonyMatrix.html">HarmonyMatrix()</a></code> function will scale expression data, run PCA, and run the Harmony integration algorithm.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(harmony)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">my_harmony_embeddings &lt;-<span class="st"> </span><span class="kw"><a href="../reference/HarmonyMatrix.html">HarmonyMatrix</a></span>(</a>
<a class="sourceLine" id="cb16-3" data-line-number="3">  <span class="dt">data_mat  =</span> normalized_counts,</a>
<a class="sourceLine" id="cb16-4" data-line-number="4">  <span class="dt">meta_data =</span> meta_data,</a>
<a class="sourceLine" id="cb16-5" data-line-number="5">  <span class="dt">vars_use  =</span> <span class="st">"dataset"</span></a>
<a class="sourceLine" id="cb16-6" data-line-number="6">)</a></code></pre></div>
</div>
<div id="seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#seurat" class="anchor"></a>Seurat</h2>
<p>You can run Harmony within your Seurat workflow. You’ll only need to make two changes to your code.</p>
<ol style="list-style-type: decimal">
<li>Run Harmony with the <code><a href="../reference/RunHarmony.html">RunHarmony()</a></code> function.</li>
<li>In downstream analyses, use the Harmony embeddings instead of PCA.</li>
</ol>
<p>For example, run Harmony and then UMAP in two lines.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">seuratObj &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunHarmony.html">RunHarmony</a></span>(seuratObj, <span class="st">"dataset"</span>)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">seuratObj &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(seuratObj, <span class="dt">reduction =</span> <span class="st">"harmony"</span>)</a></code></pre></div>
<p>For details, check out these vignettes:</p>
<ul>
<li><a href="http://htmlpreview.github.io/?https://github.com/immunogenomics/harmony/blob/master/docs/SeuratV2.html">Seurat V2</a></li>
<li><a href="http://htmlpreview.github.io/?https://github.com/immunogenomics/harmony/blob/master/docs/SeuratV3.html">Seurat V3</a></li>
</ul>
</div>
<div id="mudan" class="section level2">
<h2 class="hasAnchor">
<a href="#mudan" class="anchor"></a>MUDAN</h2>
<p>You can run Harmony with functions from the <a href="https://github.com/jefworks/mudan">MUDAN</a> R package. For more, details, check out this <a href="http://htmlpreview.github.io/?https://github.com/immunogenomics/harmony/blob/master/docs/mudan.html">vignette</a>.</p>
</div>
<div id="harmony-with-two-or-more-covariates" class="section level2">
<h2 class="hasAnchor">
<a href="#harmony-with-two-or-more-covariates" class="anchor"></a>Harmony with two or more covariates</h2>
<p>Harmony can integrate over multiple covariates. To do this, specify a vector covariates to integrate.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">harmonized_pcs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/HarmonyMatrix.html">HarmonyMatrix</a></span>(</a>
<a class="sourceLine" id="cb18-2" data-line-number="2">  <span class="dt">data_mat  =</span> scaled_pcs,</a>
<a class="sourceLine" id="cb18-3" data-line-number="3">  <span class="dt">meta_data =</span> meta_data,</a>
<a class="sourceLine" id="cb18-4" data-line-number="4">  <span class="dt">vars_use  =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dataset"</span>, <span class="st">"donor"</span>, <span class="st">"batch_id"</span>), <span class="co"># multiple covariates</span></a>
<a class="sourceLine" id="cb18-5" data-line-number="5">  <span class="dt">do_pca    =</span> <span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb18-6" data-line-number="6">)</a></code></pre></div>
<p>Do the same with your Seurat object:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">seuratObject &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunHarmony.html">RunHarmony</a></span>(seuratObject, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dataset"</span>, <span class="st">"donor"</span>, <span class="st">"batch_id"</span>))</a></code></pre></div>
</div>
<div id="detailed-breakdown-of-the-harmony-algorithm" class="section level2">
<h2 class="hasAnchor">
<a href="#detailed-breakdown-of-the-harmony-algorithm" class="anchor"></a>Detailed breakdown of the Harmony algorithm</h2>
<p>For more details on how each part of Harmony works, consult our more detailed <a href="software.broadinstitute.org/harmony/vignettes/walkthrough.html">vignette</a> “Detailed Walkthrough of Harmony Algorithm”.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</a></code></pre></div>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] cowplot_1.0.0    harmony_1.0      Rcpp_1.0.2       ggplot2_3.2.1   
## [5] dplyr_0.8.3      BiocStyle_2.10.0
## 
## loaded via a namespace (and not attached):
##  [1] compiler_3.6.0     pillar_1.4.2       BiocManager_1.30.4
##  [4] tools_3.6.0        digest_0.6.20      evaluate_0.14     
##  [7] memoise_1.1.0      tibble_2.1.3       gtable_0.3.0      
## [10] pkgconfig_2.0.2    rlang_0.4.0        yaml_2.2.0        
## [13] pkgdown_1.4.0.9000 xfun_0.9           withr_2.1.2       
## [16] stringr_1.4.0      knitr_1.24         desc_1.2.0        
## [19] fs_1.3.1           rprojroot_1.3-2    grid_3.6.0        
## [22] tidyselect_0.2.5   glue_1.3.1         R6_2.4.0          
## [25] rmarkdown_1.15     bookdown_0.13      tidyr_0.8.3       
## [28] purrr_0.3.2        magrittr_1.5       codetools_0.2-16  
## [31] backports_1.1.4    scales_1.0.0       htmltools_0.3.6   
## [34] MASS_7.3-51.4      assertthat_0.2.1   colorspace_1.4-1  
## [37] labeling_0.3       stringi_1.4.3      lazyeval_0.2.2    
## [40] munsell_0.5.0      crayon_1.3.4</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#integrating-cell-line-datasets-from-10x">Integrating cell line datasets from 10X</a></li>
      <li>
<a href="#next-steps">Next Steps</a><ul class="nav nav-pills nav-stacked">
<li><a href="#interfacing-to-software-packages">Interfacing to software packages</a></li>
      <li><a href="#normalized-gene-matrix">Normalized gene matrix</a></li>
      <li><a href="#seurat">Seurat</a></li>
      <li><a href="#mudan">MUDAN</a></li>
      <li><a href="#harmony-with-two-or-more-covariates">Harmony with two or more covariates</a></li>
      <li><a href="#detailed-breakdown-of-the-harmony-algorithm">Detailed breakdown of the Harmony algorithm</a></li>
      </ul>
</li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/ilyakorsunsky">Ilya Korsunsky</a>, Nghia Millard, <a href="https://immunogenomics.hms.harvard.edu">Soumya Raychaudhuri</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
