<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">


<title>Main Harmony interface — HarmonyMatrix • harmony</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- Bootstrap -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha256-YLGeXaapI0/5IgZopewRJcFXomhRMlYYjugPLSyNjTY=" crossorigin="anonymous" />


<!-- Bootstrap 4 JS -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.slim.min.js" integrity="sha256-pasqAKBDmFT4eHoN2ndd6lN370kFiGUFyTiUHWhU7k8=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha256-CjSoeELFOcH0/uxWu6mC/Vlrc1AARqbm/jiiImDGV3s=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.css"/>
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Main Harmony interface — HarmonyMatrix" />
<meta property="og:description" content="Use this to run the Harmony algorithm on gene expression or PCA matrix." />
<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>



  </head>

  <body data-spy="scroll" data-target="#toc">
    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light">
  <div class="container">
    <a class="navbar-brand mb-0 h1" href="../index.html">harmony</a>

    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <button type="button" class="badge badge-default d-none d-lg-block" data-toggle="tooltip" data-placement="bottom" title="" data-original-title="Released version">1.0</button>

    <div id="navbar" class="collapse navbar-collapse">
      <ul class="navbar-nav mr-auto ml-3">
      <li class="nav-item">
  <a class="nav-link" href="../index.html">Home</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/quickstart.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="navbarDropdown">
    <a class="dropdown-item" href="../advanced.html">Detailed Walkthrough</a>
    <a class="dropdown-item" href="../mudan.html">MUDAN</a>
    <a class="dropdown-item" href="../SeuratV2.html">Seurat V2</a>
    <a class="dropdown-item" href="../SeuratV3.html">Seurat V3</a>
  </div>
</li>
      </ul>

      

      <ul class="navbar-nav ml-1">
      <li class="nav-item">
  <a class="nav-link" href="https://github.com/immunogenomics/harmony">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div>
  </div>
</nav>

    <div class="container template-reference-topic">

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header pb-2 mt-4 mb-2 border-bottom">
    <h1>Main Harmony interface</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunogenomics/harmony/blob/master/R/ui.R'><code>R/ui.R</code></a></small>
    <div class="d-none name"><code>HarmonyMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use this to run the Harmony algorithm on gene expression or PCA matrix.</p>
    </div>

    <pre class="usage"><span class='fu'>HarmonyMatrix</span>(<span class='no'>data_mat</span>, <span class='no'>meta_data</span>, <span class='no'>vars_use</span>, <span class='kw'>do_pca</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>npcs</span> <span class='kw'>=</span> <span class='fl'>20</span>,
  <span class='kw'>theta</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sigma</span> <span class='kw'>=</span> <span class='fl'>0.1</span>, <span class='kw'>nclust</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>tau</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>block.size</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>max.iter.harmony</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>max.iter.cluster</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>epsilon.cluster</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>,
  <span class='kw'>epsilon.harmony</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>, <span class='kw'>plot_convergence</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>return_object</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>reference_values</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>cluster_prior</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>

    <h2>Arguments</h2>
    <dl class="row" id="ref-arguments">
      <dt class="col-sm-3" id="ref-arguments name">data_mat</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Matrix of normalized gene expession (default) or PCA 
embeddings (see do_pca). 
Cells can be rows or columns.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">meta_data</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Either (1) Dataframe with variables to integrate or (2) 
vector with labels.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">vars_use</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>If meta_data is dataframe, this defined which variable(s) 
to remove (character vector).</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">do_pca</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Whether to perform PCA on input matrix.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">npcs</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>If doing PCA on input matrix, number of PCs to compute.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">theta</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Diversity clustering penalty parameter. Specify for each
variable in vars_use Default theta=2. theta=0 does not encourage any 
diversity. Larger values of theta result in more diverse clusters.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">lambda</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Ridge regression penalty parameter. Specify for each variable
 in vars_use. 
Default lambda=1. Lambda must be strictly positive. Smaller values result 
in more aggressive correction.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">sigma</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Width of soft kmeans clusters. Default sigma=0.1. Sigma scales
the distance from a cell to cluster centroids. Larger values of sigma 
result in cells assigned to more clusters. Smaller values of sigma make 
soft kmeans cluster approach hard clustering.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">nclust</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Number of clusters in model. nclust=1 equivalent to simple 
linear regression.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">tau</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Protection against overclustering small datasets with large ones.
tau is the expected number of cells per cluster.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">block.size</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>What proportion of cells to update during clustering.
Between 0 to 1, default 0.05. Larger values may be faster but less accurate</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">max.iter.harmony</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Maximum number of rounds to run Harmony. One round
of Harmony involves one clustering and one correction step.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">max.iter.cluster</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Maximum number of rounds to run clustering at each 
round of Harmony.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">epsilon.cluster</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Convergence tolerance for clustering round of 
Harmony. Set to -Inf to never stop early.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">epsilon.harmony</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Convergence tolerance for Harmony. Set to -Inf to
never stop early.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">plot_convergence</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Whether to print the convergence plot of the 
clustering objective function. TRUE to plot, FALSE to suppress. This can be
 useful for debugging.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">return_object</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>(Advanced Usage) Whether to return the Harmony object 
or only the corrected PCA embeddings.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">verbose</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>Whether to print progress messages. TRUE to print, 
FALSE to suppress.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">reference_values</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>(Advanced Usage) Defines reference dataset(s). 
Cells that have batch variables values matching reference_values will not 
be moved.</p></dd>
      <dt class="col-sm-3" id="ref-arguments name">cluster_prior</dt>
      <dd class="col-sm-9" id="ref-arguments desc"><p>(Advanced Usage) Provides user defined clusters for 
cluster initialization. If the number of provided clusters C is less than K, 
Harmony will initialize K-C clusters with kmeans. C cannot exceed K.</p></dd>
    </dl>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>By default, matrix with corrected PCA embeddings. If return_object 
is TRUE, returns the full Harmony object (R6 reference class type).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>

## By default, Harmony inputs a normalized gene expression matrix
</div><span class='co'># NOT RUN {</span>
<span class='no'>harmony_embeddings</span> <span class='kw'>&lt;-</span> <span class='fu'>HarmonyMatrix</span>(<span class='no'>exprs_matrix</span>, <span class='no'>meta_data</span>, <span class='st'>'dataset'</span>)
<span class='co'># }</span><div class='input'>
<span class='co'>## Harmony can also take a PCA embeddings matrix</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>cell_lines_small</span>)
<span class='no'>pca_matrix</span> <span class='kw'>&lt;-</span> <span class='no'>cell_lines_small</span>$<span class='no'>scaled_pcs</span>
<span class='no'>meta_data</span> <span class='kw'>&lt;-</span> <span class='no'>cell_lines_small</span>$<span class='no'>meta_data</span>
<span class='no'>harmony_embeddings</span> <span class='kw'>&lt;-</span> <span class='fu'>HarmonyMatrix</span>(<span class='no'>pca_matrix</span>, <span class='no'>meta_data</span>, <span class='st'>'dataset'</span>,
                                    <span class='kw'>do_pca</span><span class='kw'>=</span><span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; <span class='message'>Harmony 1/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 2/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 3/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 4/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 5/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 6/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 7/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 8/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony converged after 8 iterations</span></div><div class='input'>
<span class='co'>## Output is a matrix of corrected PC embeddings</span>
<span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span>(<span class='no'>harmony_embeddings</span>)</div><div class='output co'>#&gt; [1] 300  20</div><div class='input'><span class='no'>harmony_embeddings</span>[<span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span>(<span class='fl'>5</span>), <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span>(<span class='fl'>5</span>)]</div><div class='output co'>#&gt;                X1            X2            X3            X4            X5
#&gt; [1,]  0.008448274 -0.0018134832 -1.552202e-03 -0.0077926057  0.0001466930
#&gt; [2,] -0.011292669  0.0003581516 -1.989622e-05 -0.0014619744  0.0034635288
#&gt; [3,]  0.010883023 -0.0050407453  1.579927e-03 -0.0043506676 -0.0006763035
#&gt; [4,]  0.009927078 -0.0014537998  1.159716e-03 -0.0080219594 -0.0014029851
#&gt; [5,] -0.010156191  0.0027586043 -3.074632e-03 -0.0000108491  0.0009655249</div><div class='input'>
<span class='co'>## Finally, we can return an object with all the underlying data structures</span>
<span class='no'>harmony_object</span> <span class='kw'>&lt;-</span> <span class='fu'>HarmonyMatrix</span>(<span class='no'>pca_matrix</span>, <span class='no'>meta_data</span>, <span class='st'>'dataset'</span>,
                                    <span class='kw'>do_pca</span><span class='kw'>=</span><span class='fl'>FALSE</span>, <span class='kw'>return_object</span><span class='kw'>=</span><span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; <span class='message'>Harmony 1/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 2/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 3/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 4/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 5/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 6/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 7/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 8/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 9/10</span></div><div class='output co'>#&gt; <span class='message'>Harmony 10/10</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span>(<span class='no'>harmony_object</span>$<span class='no'>Y</span>) <span class='co'>## cluster centroids</span></div><div class='output co'>#&gt; [1] 20 10</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span>(<span class='no'>harmony_object</span>$<span class='no'>R</span>) <span class='co'>## soft cluster assignment</span></div><div class='output co'>#&gt; [1]  10 300</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span>(<span class='no'>harmony_object</span>$<span class='no'>Z_corr</span>) <span class='co'>## corrected PCA embeddings</span></div><div class='output co'>#&gt; [1]  20 300</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span>(<span class='no'>harmony_object</span>$<span class='no'>O</span>) <span class='co'>## batch by cluster co-occurence matrix</span></div><div class='output co'>#&gt;           [,1]         [,2]         [,3]
#&gt; [1,]  6.197592 9.983329e+00 8.725975e-05
#&gt; [2,] 11.953170 7.462688e-06 2.536973e+01
#&gt; [3,]  7.495340 1.563652e+01 6.057042e-07
#&gt; [4,] 10.170492 1.752684e+01 7.977053e-07
#&gt; [5,] 10.999486 1.901373e-05 2.325515e+01
#&gt; [6,] 10.060248 1.894319e+01 1.766110e-06</div><div class='input'>
</div></pre>
  </div>

  <div class="col-md-3 d-none d-md-block" id="sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/ilyakorsunsky'>Ilya Korsunsky</a>, Nghia Millard, <a href='https://immunogenomics.hms.harvard.edu'>Soumya Raychaudhuri</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


