<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Harmony single cell integration — RunHarmony • harmony</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />
<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Harmony single cell integration — RunHarmony" />

<meta property="og:description" content="Run Harmony algorithm with Seurat and SingleCellAnalysis pipelines." />

<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">harmony</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/quickstart.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../advanced.html">Advanced Concepts</a>
    </li>
    <li>
      <a href="../mudan.html">MUDAN</a>
    </li>
    <li>
      <a href="../SeuratV2.html">Seurat V2</a>
    </li>
    <li>
      <a href="../SeuratV3.html">Seurat V3</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/immunogenomics/harmony">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Harmony single cell integration</h1>
    <small class="dont-index">Source: <a href='https://github.com/immunogenomics/harmony/blob/master/R/RunHarmony.R'><code>R/RunHarmony.R</code></a></small>
    <div class="hidden name"><code>RunHarmony.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Run Harmony algorithm with Seurat and SingleCellAnalysis pipelines.</p>
    
    </div>

    <pre class="usage"><span class='fu'>RunHarmony</span>(<span class='no'>object</span>, <span class='no'>group.by.vars</span>, <span class='no'>...</span>)

<span class='co'># S3 method for seurat</span>
<span class='fu'>RunHarmony</span>(<span class='no'>object</span>, <span class='no'>group.by.vars</span>, <span class='kw'>dims.use</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>theta</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sigma</span> <span class='kw'>=</span> <span class='fl'>0.1</span>, <span class='kw'>nclust</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>tau</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>block.size</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>max.iter.harmony</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>max.iter.cluster</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>epsilon.cluster</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>,
  <span class='kw'>epsilon.harmony</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>, <span class='kw'>plot_convergence</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>reference_values</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>reduction.save</span> <span class='kw'>=</span> <span class='st'>"harmony"</span>, <span class='no'>...</span>)

<span class='co'># S3 method for Seurat</span>
<span class='fu'>RunHarmony</span>(<span class='no'>object</span>, <span class='no'>group.by.vars</span>, <span class='kw'>reduction</span> <span class='kw'>=</span> <span class='st'>"pca"</span>,
  <span class='kw'>dims.use</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>theta</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sigma</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>nclust</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>tau</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>block.size</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>max.iter.harmony</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>max.iter.cluster</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>epsilon.cluster</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>,
  <span class='kw'>epsilon.harmony</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>, <span class='kw'>plot_convergence</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>reference_values</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>reduction.save</span> <span class='kw'>=</span> <span class='st'>"harmony"</span>,
  <span class='kw'>assay.use</span> <span class='kw'>=</span> <span class='st'>"RNA"</span>, <span class='kw'>project.dim</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='no'>...</span>)

<span class='co'># S3 method for SingleCellExperiment</span>
<span class='fu'>RunHarmony</span>(<span class='no'>object</span>, <span class='no'>group.by.vars</span>,
  <span class='kw'>dims.use</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>theta</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sigma</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>nclust</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>tau</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>block.size</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>max.iter.harmony</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>max.iter.cluster</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>epsilon.cluster</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>,
  <span class='kw'>epsilon.harmony</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>, <span class='kw'>plot_convergence</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>reference_values</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>reduction.save</span> <span class='kw'>=</span> <span class='st'>"HARMONY"</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Pipeline object. Must have PCA computed.</p></td>
    </tr>
    <tr>
      <th>group.by.vars</th>
      <td><p>Which variable(s) to remove (character vector).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>other parameters</p></td>
    </tr>
    <tr>
      <th>dims.use</th>
      <td><p>Which PCA dimensions to use for Harmony. By default, use all</p></td>
    </tr>
    <tr>
      <th>theta</th>
      <td><p>Diversity clustering penalty parameter. Specify for each 
variable in group.by.vars. Default theta=2. theta=0 does not encourage any
 diversity. Larger values of theta result in more diverse clusters.</p></td>
    </tr>
    <tr>
      <th>lambda</th>
      <td><p>Ridge regression penalty parameter. Specify for each variable 
in group.by.vars. Default lambda=1. Lambda must be strictly positive. 
Smaller values result in more aggressive correction.</p></td>
    </tr>
    <tr>
      <th>sigma</th>
      <td><p>Width of soft kmeans clusters. Default sigma=0.1. Sigma scales 
the distance from a cell to cluster centroids. Larger values of sigma result
 in cells assigned to more clusters. Smaller values of sigma make soft 
 kmeans cluster approach hard clustering.</p></td>
    </tr>
    <tr>
      <th>nclust</th>
      <td><p>Number of clusters in model. nclust=1 equivalent to simple
linear regression.</p></td>
    </tr>
    <tr>
      <th>tau</th>
      <td><p>Protection against overclustering small datasets with large ones.
tau is the expected number of cells per cluster.</p></td>
    </tr>
    <tr>
      <th>block.size</th>
      <td><p>What proportion of cells to update during clustering.
Between 0 to 1, default 0.05. Larger values may be faster but less accurate</p></td>
    </tr>
    <tr>
      <th>max.iter.harmony</th>
      <td><p>Maximum number of rounds to run Harmony. One round
of Harmony involves one clustering and one correction step.</p></td>
    </tr>
    <tr>
      <th>max.iter.cluster</th>
      <td><p>Maximum number of rounds to run clustering at each 
round of Harmony.</p></td>
    </tr>
    <tr>
      <th>epsilon.cluster</th>
      <td><p>Convergence tolerance for clustering round of Harmony
Set to -Inf to never stop early.</p></td>
    </tr>
    <tr>
      <th>epsilon.harmony</th>
      <td><p>Convergence tolerance for Harmony. Set to -Inf to
never stop early.</p></td>
    </tr>
    <tr>
      <th>plot_convergence</th>
      <td><p>Whether to print the convergence plot of the 
clustering objective function. TRUE to plot, FALSE to suppress. This can be
 useful for debugging.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Whether to print progress messages. TRUE to print, FALSE to
suppress.</p></td>
    </tr>
    <tr>
      <th>reference_values</th>
      <td><p>(Advanced Usage) Defines reference dataset(s). Cells
that have batch variables values matching reference_values will not be moved</p></td>
    </tr>
    <tr>
      <th>reduction.save</th>
      <td><p>Keyword to save Harmony reduction. Useful if you want
to try Harmony with multiple parameters and save them as e.g. 
'harmony_theta0', 'harmony_theta1', 'harmony_theta2'</p></td>
    </tr>
    <tr>
      <th>reduction</th>
      <td><p>Name of dimension reduction to use. Default is PCA.</p></td>
    </tr>
    <tr>
      <th>assay.use</th>
      <td><p>(Seurat V3 only) Which assay to Harmonize with (RNA 
by default).</p></td>
    </tr>
    <tr>
      <th>project.dim</th>
      <td><p>Project dimension reduction loadings. Default TRUE.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Seurat (version 2) object. Harmony dimensions placed into 
dimensional reduction object harmony. For downstream Seurat analyses, 
use reduction.use='harmony' and reduction.type='harmony'.</p>
<p>Seurat (version 3) object. Harmony dimensions placed into
dimensional reduction object harmony. For downstream Seurat analyses,
use reduction='harmony'.</p>
<p>SingleCellExperiment object. After running RunHarmony, the corrected
cell embeddings can be accessed with reducedDim(object, "Harmony").</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>## Seurat Version 2</span>
<span class='kw'>if</span> (<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/ns-load'>requireNamespace</a></span>(<span class='st'>"Seurat"</span>, <span class='kw'>quietly</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)) {
    <span class='no'>pkg_version</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/packageDescription'>packageVersion</a></span>(<span class='st'>'Seurat'</span>)
    <span class='kw'>if</span> (<span class='no'>pkg_version</span> <span class='kw'>&gt;=</span> <span class='st'>"2.0"</span> <span class='kw'>&amp;</span> <span class='no'>pkg_version</span> <span class='kw'>&lt;</span> <span class='st'>"3.0"</span>) {
        <span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>cell_lines_small_seurat_v2</span>)
        <span class='no'>seuratObject</span> <span class='kw'>&lt;-</span> <span class='fu'>RunHarmony</span>(<span class='no'>cell_lines_small_seurat_v2</span>, <span class='st'>'dataset'</span>,
                                   <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>.1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

        <span class='co'>## Harmony cell embeddings</span>
        <span class='no'>harmony_embedding</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'>GetCellEmbeddings</span>(
            <span class='no'>seuratObject</span>, <span class='st'>'harmony'</span>
        )
        <span class='no'>harmony_embedding</span>[<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>)]

        <span class='co'>## Harmony gene loadings</span>
        <span class='no'>harmony_loadings</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'>GetGeneLoadings</span>(
            <span class='no'>seuratObject</span>, <span class='st'>'harmony'</span>
        )
        <span class='no'>harmony_loadings</span>[<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>)]

        <span class='no'>p1</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/Seurat/topics/DimPlot'>DimPlot</a></span>(<span class='no'>seuratObject</span>, <span class='kw'>reduction.use</span> <span class='kw'>=</span> <span class='st'>'harmony'</span>,
                      <span class='kw'>group.by</span> <span class='kw'>=</span> <span class='st'>'dataset'</span>, <span class='kw'>do.return</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
        <span class='no'>p2</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/Seurat/topics/VlnPlot'>VlnPlot</a></span>(<span class='no'>seuratObject</span>, <span class='kw'>features.plot</span> <span class='kw'>=</span> <span class='st'>'Harmony1'</span>,
                      <span class='kw'>group.by</span> <span class='kw'>=</span> <span class='st'>'dataset'</span>, <span class='kw'>do.return</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
        <span class='kw pkg'>cowplot</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/cowplot/topics/plot_grid'>plot_grid</a></span>(<span class='no'>p1</span>,<span class='no'>p2</span>)
    }
}

<span class='co'>## Seurat Version 3</span>
<span class='kw'>if</span> (<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/ns-load'>requireNamespace</a></span>(<span class='st'>"Seurat"</span>, <span class='kw'>quietly</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)) {
    <span class='no'>pkg_version</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/packageDescription'>packageVersion</a></span>(<span class='st'>'Seurat'</span>)
    <span class='kw'>if</span> (<span class='no'>pkg_version</span> <span class='kw'>&gt;=</span> <span class='st'>"3.0"</span> <span class='kw'>&amp;</span> <span class='no'>pkg_version</span> <span class='kw'>&lt;</span> <span class='st'>"4.0"</span>) {
        <span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>cell_lines_small_seurat_v3</span>)
        <span class='no'>seuratObject</span> <span class='kw'>&lt;-</span> <span class='fu'>RunHarmony</span>(<span class='no'>cell_lines_small_seurat_v3</span>, <span class='st'>'dataset'</span>,
                                   <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>.1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
        <span class='co'>## Harmony cell embeddings</span>
        <span class='no'>harmony_embedding</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/Seurat/topics/Embeddings'>Embeddings</a></span>(<span class='no'>seuratObject</span>, <span class='st'>'harmony'</span>)
        <span class='no'>harmony_embedding</span>[<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>)]
        <span class='co'>## Harmony gene loadings</span>
        <span class='no'>harmony_loadings</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/Seurat/topics/Loadings'>Loadings</a></span>(<span class='no'>seuratObject</span>, <span class='st'>'harmony'</span>)
        <span class='no'>harmony_loadings</span>[<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>)]

        <span class='no'>p1</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/Seurat/topics/DimPlot'>DimPlot</a></span>(<span class='no'>seuratObject</span>, <span class='kw'>reduction</span> <span class='kw'>=</span> <span class='st'>'harmony'</span>,
                              <span class='kw'>group.by</span> <span class='kw'>=</span> <span class='st'>'dataset'</span>, <span class='kw'>do.return</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
        <span class='no'>p2</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/Seurat/topics/VlnPlot'>VlnPlot</a></span>(<span class='no'>seuratObject</span>, <span class='kw'>features</span> <span class='kw'>=</span> <span class='st'>'harmony_1'</span>,
                              <span class='kw'>group.by</span> <span class='kw'>=</span> <span class='st'>'dataset'</span>, <span class='kw'>do.return</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
        <span class='kw pkg'>cowplot</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/cowplot/topics/plot_grid'>plot_grid</a></span>(<span class='no'>p1</span>, <span class='no'>p2</span>)
    }
}</div><div class='output co'>#&gt; <span class='message'>The following functions and any applicable methods accept the dots: CombinePlots</span></div><div class='img'><img src='RunHarmony-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## SingleCellExperiment </span>
<span class='kw'>if</span> (<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/ns-load'>requireNamespace</a></span>(<span class='st'>"SingleCellExperiment"</span>, <span class='kw'>quietly</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)) {

    <span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>cell_lines_small_sce</span>)
    <span class='no'>sceObject</span> <span class='kw'>&lt;-</span> <span class='fu'>RunHarmony</span>(<span class='no'>cell_lines_small_sce</span>, <span class='st'>'dataset'</span>,
                            <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>.1</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

    <span class='co'>## Harmony cell embeddings</span>
    <span class='no'>harmony_embedding</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>SingleCellExperiment</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/SingleCellExperiment/topics/SCE-reducedDims'>reducedDim</a></span>(
        <span class='no'>sceObject</span>, <span class='st'>'HARMONY'</span>
    )
    <span class='no'>harmony_embedding</span>[<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>), <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/seq'>seq_len</a></span>(<span class='fl'>5</span>)]

    <span class='co'>## Plot the Harmonized embeddings</span>
    <span class='co'>## Colored by batch and cell type</span>
    <span class='kw pkg'>SingleCellExperiment</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/SingleCellExperiment/topics/SCE-reducedDims'>reducedDim</a></span>(<span class='no'>sceObject</span>, <span class='st'>'HARMONY'</span>) <span class='kw'>%&gt;%</span>
        <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/cbind'>cbind</a></span>(<span class='kw pkg'>SingleCellExperiment</span><span class='kw ns'>::</span><span class='fu'>colData</span>(<span class='no'>sceObject</span>)) <span class='kw'>%&gt;%</span>
        <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/data.frame'>data.frame</a></span>() <span class='kw'>%&gt;%</span>
    <span class='kw pkg'>tidyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://tidyr.tidyverse.org/reference/gather.html'>gather</a></span>(<span class='no'>key</span>, <span class='no'>val</span>, <span class='no'>dataset</span>, <span class='no'>cell_type</span>) <span class='kw'>%&gt;%</span>
    <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/mutate.html'>mutate</a></span>(<span class='kw'>key</span> <span class='kw'>=</span> <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/case_when.html'>case_when</a></span>(
        <span class='no'>key</span> <span class='kw'>==</span> <span class='st'>'dataset'</span> ~ <span class='st'>'Dataset batch'</span>,
        <span class='no'>key</span> <span class='kw'>==</span> <span class='st'>'cell_type'</span> ~ <span class='st'>'Known cell types'</span>
    )) <span class='kw'>%&gt;%</span>
    <span class='kw pkg'>dplyr</span><span class='kw ns'>::</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/sample.html'>sample_frac</a></span>(<span class='fl'>1L</span>, <span class='fl'>FALSE</span>) <span class='kw'>%&gt;%</span>
    <span class='kw pkg'>ggplot2</span><span class='kw ns'>::</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span>(<span class='kw pkg'>ggplot2</span><span class='kw ns'>::</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span>(<span class='kw'>x</span> <span class='kw'>=</span> <span class='no'>harmony_1</span>,
                                 <span class='kw'>y</span> <span class='kw'>=</span> <span class='no'>harmony_2</span>,
                                 <span class='kw'>color</span> <span class='kw'>=</span> <span class='no'>val</span>)) +
        <span class='kw pkg'>ggplot2</span><span class='kw ns'>::</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_point.html'>geom_point</a></span>() +
        <span class='kw pkg'>ggplot2</span><span class='kw ns'>::</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/facet_wrap.html'>facet_wrap</a></span>(~<span class='no'>key</span>) +
        <span class='kw pkg'>ggplot2</span><span class='kw ns'>::</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggtheme.html'>theme_test</a></span>(<span class='kw'>base_size</span> <span class='kw'>=</span> <span class='fl'>12</span>) +
        <span class='kw pkg'>ggplot2</span><span class='kw ns'>::</span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>labs</a></span>(<span class='kw'>title</span> <span class='kw'>=</span> <span class='st'>'Cell embeddings after Harmony'</span>)
}</div><div class='img'><img src='RunHarmony-2.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/ilyakorsunsky'>Ilya Korsunsky</a>, Nghia Millard, <a href='https://immunogenomics.hms.harvard.edu'>Soumya Raychaudhuri</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

